Login: Add missing support for 2FA codes with white spaces

stephanethomas opened the pull request:
This pull request is a follow-up of https://github.com/Automattic/wp-calypso/pull/23042 that updates input fields with codes (backup, recovery, sms, and app codes) in order to accept codes formatted with white spaces for readability purpose. It does so by introducing a new `FormVerificationCodeInput` component.
 
### Testing instructions

Note you'll have to patch your sandbox with D11117-code if it has not been merged yet.

#### Default login

1. Run `git checkout update/verification-code-fields` and start your server, or open a [live branch](https://calypso.live/?branch=update/verification-code-fields)

---

##### Two-Step Authentication (2FA) via SMS

###### Enabling 2FA via SMS

2. Open the [`Login` page](http://calypso.localhost:3000/log-in) in an incognito window
3. Sign in with an account with 2FA disabled
4. Open the [`Two-Step Authentication` page](http://calypso.localhost:3000/me/security/two-step)
5. Click the `Get Started` button
6. Enter a country code and a phone number
7. Click the `Verify via SMS` button
8. Assert that the code you receive matches the format of the placeholder of the `Enter the code you receive via SMS` field
9. Enter the code with white spaces in this field, and click the `Enable` button
10. Take note of backup codes
11. Check the `I have printed or saved these codes` checkbox, and click the `All Finished!` button
12. Assert that a `Successfully enabled Two-Step Authentication.` message is displayed
13. Assert that two-step authentication is effectively enabled by reloading the page

###### Verifying a 2FA backup code

14. Locate the `Type a Backup Code to Verify` field on the `Two-Step Authentication` page
15. Assert that the placeholder of this field matches the format of the backup code
16. Enter a backup code, and click the `Verify` button
17. Assert that a `Backup codes have been verified` message is now displayed
18. Assert that this is still the case when reloading the page

###### Signing in with 2FA via SMS

19. Open the [`Login` page](http://calypso.localhost:3000/log-in) in an incognito window
20. Submit credentials of the account just used above
21. Assert that the code you receive matches the format of the placeholder of the `Verification code` field
22. Enter the code with white spaces in this field, and click the `Continue` button
23. Assert that you can log in to WordPress.com with this code

###### Reauthenticating with 2FA via SMS

24. Delete the `twostep_auth` cookie set on the `public-api.wordpress.com` domain
25. Navigate to the [`Me` section](http://calypso.localhost:3000/me)
26. Assert that you are presented with a modal asking you to authenticate again
27. Click the `Sens SMS Code` button
28. Assert that the code you receive matches the format of the placeholder of the `Verification Code` field
29. Enter the code with white spaces in this field, and click the `Verify` button
30. Assert that you can now access the `My Profile` page

###### Disabling 2FA via SMS

31. Click the `Disable Two-Step Authentication` button on the `Two-Step Authentication` page
32. Assert that the code you receive matches the format of the placeholder of the `Verification Code` field
33. Enter the code with white spaces in this field, and click the `Disable Two-Step` button
34. Assert that a `Successfully disabled Two-Step Authentication.` message is displayed
35. Assert that two-step authentication is effectively disabled by reloading the page

---

##### Two-Step Authentication (2FA) via Auth app

###### Enabling 2FA via Auth app

2. Open the [`Login` page](http://calypso.localhost:3000/log-in) in an incognito window
3. Sign in with an account with 2FA disabled
4. Open the [`Two-Step Authentication` page](http://calypso.localhost:3000/me/security/two-step)
5. Click the `Get Started` button
6. Enter a phone number, and click the `Verify via App` button
7. Scan the QR code with an Authenticator app
8. Assert that the code generated by the app matches the format of the placeholder of the input field located below the QR code
9. Enter the code with white spaces in this field, and click the `Enable` button
10. Take note of backup codes
11. Check the `I have printed or saved these codes` checkbox, and click the `All Finished!` button
12. Assert that a `Successfully enabled Two-Step Authentication.` message is displayed
13. Assert that two-step authentication is effectively enabled by reloading the page

###### Disabling 2FA via Auth app

14. Click the `Disable Two-Step Authentication` button on the `Two-Step Authentication` page
15. Assert that the code generated by your Authenticator app matches the format of the placeholder of the `Verification Code` field
16. Enter the code with white spaces in this field, and click the `Disable Two-Step` button
17. Assert that a `Successfully disabled Two-Step Authentication.` message is displayed
18. Assert that two-step authentication is effectively disabled by reloading the page

---

##### Account recovery via SMS

###### Enabling recovery via SMS

2. Open the [`Login` page](http://calypso.localhost:3000/log-in) in an incognito window
3. Sign in with an account with 2FA disabled
4. Open the [`Account Recovery` page](http://calypso.localhost:3000/me/security/account-recovery)
5. Click the `Add` button in front of `Recovery SMS Number`
6. Enter a phone number, and click the `Save Number` button
7. Assert that the code you receive matches the format of the placeholder of the `Enter the code you receive via SMS` field
8. Enter the code with white spaces in this field, and click the `Validate` button
9. Assert that a `Your recovery SMS number has been validated successfully.` message is displayed

###### Resetting password via SMS

10. Open the [`Login` page](http://calypso.localhost:3000/log-in) in an incognito window
11. Click the `Lost your password?` link in the footer
12. Enter the username or email address of the account just used above
13. Click the `Get New Password` button
14. Assert that the code you receive matches the format of the placeholder of the input field
15. Enter the code with white spaces in this field, and click the `Reset my password!` button
16. Assert that your password has been reset

###### Resetting password via SMS (new flow, not deployed)

10. Open the [`Account Recovery` page](http://calypso.localhost:3000/account-recovery) in an incognito window
11. Enter the username or email address of the account just used above
12. Click the `Get New Password` button
13. Select `Send a reset code to your main phone.`, and click the `Continue` button
14. Assert that the code you receive matches the format of the example given in the copy
15. Enter the code with white spaces in this field, and click the `Continue` button
16. Assert that your password has been reset

---

#### OAuth login

1. Make sure `config/secrets.json` defines valid `desktop_oauth_*` properties (see PCYsg-6rI-p2)
2. Run `ENABLE_FEATURES=oauth npm start` to start your server with OAuth enabled
3. Open the [`OAuth Login` page](http://calypso.localhost:3000/oauth-login) in an incognito window
4. Submit credentials for an account with 2FA enabled
5. Enter a code with white spaces in this field, and click the `Verify` button
6. Assert that you can log in to WordPress.com with this code

### To do

- [ ] Address linting issues
- [x] Rebase in to discrete commits
- [x] Fix `Stateless function components cannot be given refs.` error in `Verification Code` page
- [x] Update `Auth` and `ReauthRequired` components to allow white spaces (see D10671-code#237983)
- [x] Update [`Reset Password` page](https://wordpress.com/wp-login.php?action=sms) to allow white spaces

### Reviews
 
- [ ] Code
- [ ] Product

a8ci18n commented on client/me/security-2fa-enable/index.jsx:
```diff
@@ -328,31 +328,30 @@ class Security2faEnable extends React.Component {
 						{ this.props.translate(
-							'A code has been sent to your device via SMS.  ' +
-								'You may request another code after one minute.'
+							'A code has been sent to your device via SMS. ' +
```
Hi! I've found a possible matching string that has already been translated **26** times:
`translate( 'A code has been sent to your device via SMS.  You may request another code after one minute.' )` *ES Score: 14*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

gwwar commented:
> Address linting issues

For next time, if you see linting errors on old code (not new lines), we can skip/ignore them and fix them up in a janitorial PR. It's usually much easier to review when we break out PR with behavior changes, vs janitorial PRs.

stephanethomas commented on client/auth/login.jsx:
```diff
@@ -135,7 +136,7 @@ export class Auth extends Component {
 									<FormTextInput
 										name="auth_code"
-										type="number"
+										type="tel"
```
I switched to `tel` because `number` would not allow white spaces.

gwwar replied:
Seems odd to do so when it's not a telephone number, can we use a plain input? @ryelle this would probably read a little funny on VoiceOver wouldn't it?

stephanethomas replied:
Well, using `tel` will cause the numeric keyboard to appear on mobile devices.

ryelle replied:
It looks like screen readers should still see this as a text input, I don't think it would negatively affect accessibility. except, as mentioned, changing the keyboard -- but an auth code is going to be a number anyway, right?

In testing a `tel`, Safari did try to give me a phone number autofill menu, so maybe turn off autocomplete on this field?

gwwar replied:
Would it help if we did this somewhat similar to how we do credit card formatting? I believe that ends up being a display field and an actual raw value field. This also allows us to show the space as folks type in the auth code.

stephanethomas replied:
> but an auth code is going to be a number anyway, right?

This code is all digits separated by a white space. Entering a code without any white space is perfectly valid.

> In testing a tel, Safari did try to give me a phone number autofill menu, so maybe turn off autocomplete on this field?

Good catch, this form is missing that attribute ! I'll add it.

> I believe that ends up being a display field and an actual raw value field

Wouldn't that prevent the numeric keyboard to show up?

gwwar commented:
What does the `punted` label mean in this case? I prefer closing out PRs if we don't plan to have any activity. Explaining why we punt/close is useful too for other folks reading back.

stephanethomas commented:
> What does the punted label mean in this case? I prefer closing out PRs if we don't plan to have any activity. Explaining why we punt/close is useful too for other folks reading back.

I planned to do that, you were just too quick! 😅

stephanethomas commented:
I have addressed all remaining points, either in this pull request or the server patch D11117-code.

@catehstn, @nheagy: we're ready to enable white spaces in SMS codes everywhere, just let us know when we can carry on. Until then, I'll put this pull request in `punted`.

stephanethomas commented on client/components/forms/form-verification-code-input/index.jsx:
```diff
@@ -0,0 +1,64 @@
+	render() {
+		const { className, isError, isValid, method, ...otherProps } = this.props;
+
+		const classes = classNames( 'form-tel-verification-code', className, {
```
I think the class name should be `form-verification-code`.

stephanethomas commented:
Just rebased, addressed the [wrong class name](https://github.com/Automattic/wp-calypso/pull/23176#discussion_r176494276), and [disabled autocompletion](https://github.com/Automattic/wp-calypso/pull/23176#discussion_r176203238).

stephanethomas commented:
Just pushed a commit to update a string in the account recovery section, and rebased in the process.

gwwar commented:
@stephanethomas are we still planning to work on this? Are we blocked by anything?

If not, let's close out this PR with a note in the summary on why this was punted. If we revisit this work it's pretty easy to reopen the PR.

stephanethomas commented:
@gwwar: we're waiting for green light [from Hogwarts](https://github.com/wordpress-mobile/WordPress-iOS/pull/8984#issuecomment-379136212).

gwwar commented:
@stephanethomas 👍 let's remove the punted label then, and update the summary to note that we're waiting for a new mobile app version to release. Depending on the timeline we can still get this prepped to go on the Calypso side.

stephanethomas commented:
I'm not sure to understand, isn't the `[Status] Punted` label exactly for mentioning this sort of delay?

By the way I see we also have `[Status] Blocked/ Hold` and `[Status] Stale`. However I couldn't find any information about the exact meaning of our different labels.

gwwar commented:
I will be lobbying to remove both labels :) I think it's clearer to update the summary with any sort of status like that.

catehstn commented:
FYI @nheagy checked the stats and I think we will still need to give it a couple more weeks.

stephanethomas commented:
@catehstn, @nheagy: how are we doing now? 🙂

nheagy commented:
Looks like we're down to 10% of users with an old/broken version according to @elibud. Seems like we're good to go!

stephanethomas commented:
Just rebased, and fixed conflicts.

yurynix commented:
I've rebased & tested. I'm going to ship it.
